modules/navier_stokes/src/executioners/LinearAssemblySegregatedSolve.C

```diff
@@ -0,0 +1,459 @@
+  if (_has_passive_scalar_systems)
+    for (auto system_i : index_range(_passive_scalar_system_names))
+    {
+      _passive_scalar_system_numbers.push_back(
+          _problem.linearSysNum(_passive_scalar_system_names[system_i]));
+      _passive_scalar_systems.push_back(
+          &_problem.getLinearSystem(_passive_scalar_system_numbers[system_i]));
+      _systems_to_solve.push_back(_passive_scalar_systems.back());
+    }
+}
+
+void
+LinearAssemblySegregatedSolve::linkRhieChowUserObject()
+{
+  _rc_uo =
+      const_cast<RhieChowMassFlux *>(&getUserObject<RhieChowMassFlux>("rhie_chow_user_object"));
+  _rc_uo->linkMomentumPressureSystems(
+      _momentum_systems, _pressure_system, _momentum_system_numbers);
+
+  // Initialize the face velocities in the RC object
+  if (!_app.isRecovering())
+    _rc_uo->initFaceMassFlux();
+  _rc_uo->initCouplingField();
+}
+
+std::vector<std::pair<unsigned int, Real>>
+LinearAssemblySegregatedSolve::solveMomentumPredictor()
+{
+  // Temporary storage for the (flux-normalized) residuals from
+  // different momentum components
+  std::vector<std::pair<unsigned int, Real>> its_normalized_residuals;
+
+  LinearImplicitSystem & momentum_system_0 =
+      libMesh::cast_ref<LinearImplicitSystem &>(_momentum_systems[0]->system());
+
+  PetscLinearSolver<Real> & momentum_solver =
+      libMesh::cast_ref<PetscLinearSolver<Real> &>(*momentum_system_0.get_linear_solver());
+
+  // Solve the momentum equations.
+  // TO DO: These equations are VERY similar. If we can store the differences (things coming from
+  // BCs for example) separately, it is enough to construct one matrix.
+  for (const auto system_i : index_range(_momentum_systems))
+  {
+    _problem.setCurrentLinearSystem(_momentum_system_numbers[system_i]);
+
+    // We will need the right hand side and the solution of the next component
+    LinearImplicitSystem & momentum_system =
+        libMesh::cast_ref<LinearImplicitSystem &>(_momentum_systems[system_i]->system());
+
+    NumericVector<Number> & solution = *(momentum_system.solution);
+    NumericVector<Number> & rhs = *(momentum_system.rhs);
+    SparseMatrix<Number> & mmat = *(momentum_system.matrix);
+
+    auto diff_diagonal = solution.zero_clone();
+
+    // We assemble the matrix and the right hand side
+    _problem.computeLinearSystemSys(momentum_system, mmat, rhs);
+
+    // Still need to relax the right hand side with the same vector
+    NS::FV::relaxMatrix(mmat, _momentum_equation_relaxation, *diff_diagonal);
+    NS::FV::relaxRightHandSide(rhs, solution, *diff_diagonal);
+
+    // The normalization factor depends on the right hand side so we need to recompute it for this
+    // component
+    Real norm_factor = NS::FV::computeNormalizationFactor(solution, mmat, rhs);
+
+    // Very important, for deciding the convergence, we need the unpreconditioned
+    // norms in the linear solve
+    LIBMESH_CHKERR(KSPSetNormType(momentum_solver.ksp(), KSP_NORM_UNPRECONDITIONED));
```

Comment:
This (and the other `LIBMESH_CHKERR` calls below) just broke libMesh testing, thanks to me foolishly assuming that just because MOOSE tested clean against https://github.com/libMesh/libmesh/pull/4037 a month ago it would still be clean when I noticed the overlooked PR and hit Merge today.

Can we get them changed to LibmeshPetscCall?

Reply:
Yes, I will submit the PR in 15 mins.